test/integration/upstreams/per_host_upstream_request.h

```diff
@@ -0,0 +1,110 @@
+  void onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                     absl::string_view transport_failure_reason,
+                     Upstream::HostDescriptionConstSharedPtr host) override;
+  void onPoolReady(Envoy::Http::RequestEncoder& callbacks_encoder,
+                   Upstream::HostDescriptionConstSharedPtr host,
+                   const StreamInfo::StreamInfo& info) override;
+  Upstream::HostDescriptionConstSharedPtr host() const override { return conn_pool_->host(); }
+
+  bool valid() { return conn_pool_ != nullptr; }
+
+private:
+  // Points to the actual connection pool to create streams from.
+  Envoy::Http::ConnectionPool::Instance* conn_pool_{};
+  Envoy::Http::ConnectionPool::Cancellable* conn_pool_stream_handle_{};
+  Router::GenericConnectionPoolCallbacks* callbacks_{};
+};
+
+class PerHostHttpUpstream : public Router::GenericUpstream {
+public:
+  PerHostHttpUpstream(Router::UpstreamToDownstream& upstream_request,
+                      Envoy::Http::RequestEncoder* encoder,
+                      Upstream::HostDescriptionConstSharedPtr host)
+      : sub_upstream_(upstream_request, encoder), host_(host) {}
+
+  // GenericUpstream
+  void encodeData(Buffer::Instance& data, bool end_stream) override {
+    sub_upstream_.encodeData(data, end_stream);
+  }
+
+  void encodeMetadata(const Envoy::Http::MetadataMapVector& metadata_map_vector) override {
+    sub_upstream_.encodeMetadata(metadata_map_vector);
+  }
+
+  void encodeHeaders(const Envoy::Http::RequestHeaderMap& headers, bool end_stream) override {
+    auto dup = Envoy::Http::RequestHeaderMapImpl::create();
+    Envoy::Http::HeaderMapImpl::copyFrom(*dup, headers);
+    dup->setCopy(Envoy::Http::LowerCaseString("X-foo"), "foo-common");
+    if (auto filter_metadata = host_->cluster().metadata().filter_metadata().find("foo");
+        filter_metadata != host_->cluster().metadata().filter_metadata().end()) {
+      const ProtobufWkt::Struct& data_struct = filter_metadata->second;
+      const auto& fields = data_struct.fields();
+      if (auto iter = fields.find("bar"); iter != fields.end()) {
+        if (iter->second.kind_case() == ProtobufWkt::Value::kStringValue) {
+          dup->setCopy(Envoy::Http::LowerCaseString("X-cluster-foo"), iter->second.string_value());
+        }
+      }
+    }
+    if (host_->metadata() != nullptr) {
+      if (auto filter_metadata = host_->metadata()->filter_metadata().find("foo");
+          filter_metadata != host_->metadata()->filter_metadata().end()) {
+        const ProtobufWkt::Struct& data_struct = filter_metadata->second;
+        const auto& fields = data_struct.fields();
+        if (auto iter = fields.find("bar"); iter != fields.end()) {
+          if (iter->second.kind_case() == ProtobufWkt::Value::kStringValue) {
+            dup->setCopy(Envoy::Http::LowerCaseString("X-host-foo"), iter->second.string_value());
+          }
+        }
+      }
+    }
+    sub_upstream_.encodeHeaders(*dup, end_stream);
+  }
```

Comment:
This reads very densely, could you break it up or add comments to make it clearer what's being done?

Reply:
add a static function to inject header